### PR TITLE
[WIP] Contextual narrowing to object

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Related: https://github.com/microsoft/TypeScript/issues/48468

### CODE REVIEW

File: tests/baselines/reference/contextualNarrowingFromUnknownToObjects(strictnullchecks=true).errors.txt [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+tests/cases/compiler/contextualNarrowingFromUnknownToObjects.ts(54,31): error TS2345: Argument of type 'T' is not assignable to parameter of type '{}'.
+tests/cases/compiler/contextualNarrowingFromUnknownToObjects.ts(54,63): error TS2345: Argument of type 'T' is not assignable to parameter of type '{}'.
+
+
+==== tests/cases/compiler/contextualNarrowingFromUnknownToObjects.ts (8 errors) ====
+    declare function keysOfEmptyObject(o: {}): string[];
+    declare function keysOfNonPrimitive(o: object): string[];
+    
+    namespace explicitConstraints {
+        export function keyLengthsEqualUsingEmptyObjectFn<T>(a: T, b: T): [T, T] | undefined {
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/contextualNarrowingFromUnknownToObjects(strictnullchecks=true).errors.txt:16]: This function isn't using explicit constraints... 🤦‍♂️. I'll have to fix that.

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

**Comment by DanielRosenwasser** [comment]
Seems like there's a slowdown from this - maybe a fluke?

@typescript-bot perf test faster

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -25338,10 +25338,25 @@ namespace ts {
...
+            if (!isConstraintPosition(type, reference) && !(contextualType = tryGetContextualTypeWithNoGenericTypes(reference, checkMode))) {
+                return type;
+            }
+
+            const substituteConstraints =
+                // When we have a type parameter constrained to a union type, we can typically narrow to get better results.
+                someType(type, isGenericTypeWithUnionConstraint) ||
+                // When the contextual type is 'unknown', we may need to narrow for compatibility with non-null targets.
+                // This allows some parity with a constraint of '{} | null | undefined'.
+                (type.flags & TypeFlags.Instantiable) && contextualType && isEmptyObjectType(contextualType);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:25357]: Should this be gated on `strictNullChecks`?

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -25338,10 +25338,25 @@ namespace ts {
...
-                (isConstraintPosition(type, reference) || hasContextualTypeWithNoGenericTypes(reference, checkMode));
-            return substituteConstraints ? mapType(type, t => t.flags & TypeFlags.Instantiable ? getBaseConstraintOrType(t) : t) : type;
+            if (checkMode && checkMode & CheckMode.Inferential) {
+                return type;
+            }
+
+            let contextualType: Type | undefined;
+            // If we aren't in a constraint position, or we can't find a contextual type, or the contextual type indicates
+            // that the type in question may be a direct inference source, then don't do anything special.
+            if (!isConstraintPosition(type, reference) && !(contextualType = tryGetContextualTypeWithNoGenericTypes(reference, checkMode))) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts:25348]: I'm pretty sure the slowdown comes from fetching the contextual type way more often. Previously we'd only do it when `someType(type, isGenericTypeWithUnionConstraint)` is true, which is fairly rare.

Comment (ahejlsberg) [reply src/compiler/checker.ts:25348]: I'd suggest adding a check that `someType(type, t => !!(t.flags & TypeFlags.Instantiable))`, and bail out if not _before_ fetching the contextual type. That would reduce the overhead to only cases where something is generic.

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test faster

**Comment by DanielRosenwasser** [comment]
RWC reveals a few interesting things:

1. Substituting `unknown` for the error messages is definitely a bit confusing.
1. This gets worse with unions - e.g.

   ```ts
   let f = <T, U extends UsefulType>(x: T | U): object => x
   ```

   So whereas this used to report that `T | U` wasn't assignable to `object`, it now reports `unknown` isn't assignable to `object`.

Going to rerun with latest changes.

**Comment by ahejlsberg** [comment]
Hmm, yeah, I don't much like the error message changes in the RWC baselines. The `unknown` makes it seem like the expression doesn't have a type at all.

**Comment by DanielRosenwasser** [comment]
Discussed with @ahejlsberg offline - this PR is a bit too corner-casey and produces worse messages when a type *isn't* narrowed. We're interested in seeing if we can leverage smarter intersection types with `{}` while narrowing generics, which is a change that will have to wait for 4.8 at the least. For now we'll have to live with the break, which people can get around by adding a constraint of `{} | null | undefined`.

**Pull request closed without merging**